// @prettier/plugin-pug/dist/options/attribute-sorting/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PUG_SORT_ATTRIBUTES_OPTION = exports.PUG_SORT_ATTRIBUTES_END_OPTION = exports.PUG_SORT_ATTRIBUTES_BEGINNING_OPTION = void 0;
const __1 = require("..");
const pugSortAttributesOption = {
    since: '1.7.0',
    category: __1.CATEGORY_PUG,
    type: 'path',
    array: true,
    default: [{ value: [] }],
    description: ''
};
exports.PUG_SORT_ATTRIBUTES_BEGINNING_OPTION = {
    ...pugSortAttributesOption,
    description: 'Define a list of patterns for attributes that are sorted to the beginning.'
};
exports.PUG_SORT_ATTRIBUTES_END_OPTION = {
    ...pugSortAttributesOption,
    description: 'Define a list of patterns for attributes that are sorted at the end.'
};
exports.PUG_SORT_ATTRIBUTES_OPTION = {
    since: '1.8.0',
    category: __1.CATEGORY_PUG,
    type: 'choice',
    default: 'as-is',
    description: 'Change how the attributes between _beginning_ and _end_ should be sorted.',
    choices: [
        { value: 'asc', description: 'Sort middle attributes ascending.' },
        { value: 'desc', description: 'Sort middle attributes descending.' },
        { value: 'as-is', description: 'Middle attributes are leave untouched.' }
    ]
};
