// @prettier/plugin-pug/dist/options/empty-attributes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PUG_EMPTY_ATTRIBUTES_FORCE_QUOTES_OPTION = exports.PUG_EMPTY_ATTRIBUTES_OPTION = void 0;
const __1 = require("..");
exports.PUG_EMPTY_ATTRIBUTES_OPTION = {
    since: '1.10.0',
    category: __1.CATEGORY_PUG,
    type: 'choice',
    default: 'as-is',
    description: 'Change behavior of boolean attributes.',
    choices: [
        {
            value: 'as-is',
            description: 'Nothing is changed.'
        },
        {
            value: 'none',
            description: 'Every attribute with empty quotes will have them removed.'
        },
        {
            value: 'all',
            description: 'Every boolean attribute will be expressed with empty quotes.'
        }
    ]
};
exports.PUG_EMPTY_ATTRIBUTES_FORCE_QUOTES_OPTION = {
    since: '1.10.0',
    category: __1.CATEGORY_PUG,
    type: 'path',
    default: [{ value: [] }],
    array: true,
    description: 'Define a list of patterns for attributes that will be forced to have empty quotes even with "none" selected.'
};
