// @prettier/plugin-pug/dist/options/closing-bracket-position.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveClosingBracketPositionOption = exports.PUG_CLOSING_BRACKET_POSITION_OPTION = exports.CLOSING_BRACKET_POSITION_OPTION = void 0;
const _1 = require(".");
exports.CLOSING_BRACKET_POSITION_OPTION = {
    since: '1.3.0',
    category: _1.CATEGORY_PUG,
    type: 'choice',
    default: 'new-line',
    description: 'Determines position of closing bracket which wraps attributes.',
    choices: [
        {
            value: 'new-line',
            description: `
				Closing bracket ends with a new line.
				Example:
				input(
					type='text',
					value='my_value',
					name='my_name',
					alt='my_alt',
					autocomplete='on'
				)
				`
        },
        {
            value: 'last-line',
            description: `
			Closing bracket remains with last attribute's line.
			Example:
			input(
				type='text',
				value='my_value',
				name='my_name',
				alt='my_alt',
				autocomplete='on')
			`
        }
    ]
};
exports.PUG_CLOSING_BRACKET_POSITION_OPTION = {
    ...exports.CLOSING_BRACKET_POSITION_OPTION,
    since: '1.6.0',
    default: null,
    choices: [
        {
            value: null,
            description: 'Use `closingBracketPosition` value.'
        },
        {
            value: 'new-line',
            description: `
				Closing bracket ends with a new line.
				Example:
				input(
					type='text',
					value='my_value',
					name='my_name',
					alt='my_alt',
					autocomplete='on'
				)
				`
        },
        {
            value: 'last-line',
            description: `
			Closing bracket remains with last attribute's line.
			Example:
			input(
				type='text',
				value='my_value',
				name='my_name',
				alt='my_alt',
				autocomplete='on')
			`
        }
    ]
};
function resolveClosingBracketPositionOption(closingBracketPosition) {
    switch (closingBracketPosition) {
        case 'new-line':
            return true;
        case 'last-line':
            return false;
    }
    throw new Error(`Invalid option for pug closingBracketPosition. Found '${closingBracketPosition}'. Possible options: 'new-line' or 'last-line'`);
}
exports.resolveClosingBracketPositionOption = resolveClosingBracketPositionOption;
