// coc-lists/lib/cmdhistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const regex = /^\s*\d+\s+(.*)$/;
class Cmdhistory {
    constructor(nvim) {
        this.nvim = nvim;
        this.name = 'cmdhistory';
        this.description = 'history of vim commands';
        this.defaultAction = 'execute';
        this.actions = [];
        this.actions.push({
            name: 'execute',
            execute: async (item) => {
                if (Array.isArray(item))
                    return;
                nvim.command(item.data.cmd, true);
            }
        });
    }
    async loadItems(_context) {
        let { nvim } = this;
        let list = await nvim.eval(`split(execute("history cmd"),"\n")`);
        list = list.slice(1);
        let res = [];
        for (let line of list) {
            let ms = line.match(regex);
            if (ms) {
                let [, cmd] = ms;
                res.push({
                    label: cmd,
                    filterText: cmd,
                    data: {
                        cmd
                    }
                });
            }
        }
        return res;
    }
}
exports.default = Cmdhistory;
