// coc-lists/lib/colors.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const coc_nvim_1 = require("coc.nvim");
const path_1 = __importDefault(require("path"));
class Colors extends coc_nvim_1.BasicList {
    constructor(nvim) {
        super(nvim);
        this.name = 'colors';
        this.description = 'color schemes';
        this.defaultAction = 'set';
        this.actions = [];
        this.addLocationActions();
        this.addAction('set', item => {
            if (Array.isArray(item))
                return;
            nvim.command(`colorscheme ${item.data.name}`, true);
        });
    }
    async loadItems(_context) {
        let { nvim } = this;
        let colors = await nvim.eval('split(globpath(&rtp, "colors/*.vim"),"\n")');
        let hasPackages = await nvim.call('has', ['packages']);
        if (hasPackages) {
            let packageColors = await nvim.eval('split(globpath(&packpath, "pack/*/opt/*/colors/*.vim"),"\n")');
            colors.push(...packageColors);
        }
        return colors.map(file => {
            let name = path_1.default.basename(file, '.vim');
            return {
                label: `${name}\t${file}`,
                filterText: name,
                data: { name },
                location: coc_nvim_1.Uri.file(file).toString()
            };
        });
    }
    doHighlight() {
        let { nvim } = this;
        nvim.pauseNotification();
        nvim.command('syntax match CocColorsName /\\v^[^\\t]+/ contained containedin=CocColorsLine', true);
        nvim.command('syntax match CocColorsFile /\\t.*$/ contained containedin=CocColorsLine', true);
        nvim.command('highlight default link CocColorsName Identifier', true);
        nvim.command('highlight default link CocColorsFile Comment', true);
        nvim.resumeNotification(false, true).catch(_e => {
            // noop
        });
    }
}
exports.default = Colors;
