// coc-lists/lib/filetypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
class Filetypes {
    constructor(nvim) {
        this.nvim = nvim;
        this.name = 'filetypes';
        this.description = 'available filetypes';
        this.defaultAction = 'set';
        this.actions = [];
        this.actions.push({
            name: 'set',
            execute: item => {
                if (Array.isArray(item)) {
                    for (let i of item) {
                        nvim.command(`setf ${i.label}`, true);
                    }
                }
                else {
                    nvim.command(`setf ${item.label}`, true);
                }
            }
        });
    }
    async loadItems(_context) {
        let { nvim } = this;
        let filetypes = await nvim.eval(`sort(map(split(globpath(&rtp, 'syntax/*.vim'), '\n'),'fnamemodify(v:val, ":t:r")'))`);
        filetypes = util_1.distinct(filetypes);
        return filetypes.map(filetype => {
            return { label: filetype };
        });
    }
}
exports.default = Filetypes;
