// coc-lists/lib/marks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const safe_1 = __importDefault(require("colors/safe"));
const regex = /^\s*(\S)\s+(\d+)\s+(\d+)\s*(.*)$/;
class Marks {
    constructor(nvim) {
        this.nvim = nvim;
        this.name = 'marks';
        this.description = 'marks list';
        this.defaultAction = 'jump';
        this.actions = [];
        this.actions.push({
            name: 'jump',
            execute: item => {
                if (Array.isArray(item))
                    return;
                nvim.command('normal! `' + item.data.name + 'zz', true);
            }
        });
    }
    async loadItems(_context) {
        let { nvim } = this;
        let list = await nvim.eval('split(execute("marks"), "\n")');
        list = list.slice(1);
        list = list.filter(s => regex.test(s));
        return list.map(line => {
            let [, name, lnum, col, text] = line.match(regex);
            return {
                label: `${safe_1.default.magenta(name)}\t${text}\t${safe_1.default.grey(lnum)}\t${safe_1.default.grey(col)}`,
                filterText: name + ' ' + text,
                data: { name }
            };
        });
    }
}
exports.default = Marks;
