// coc-lists/lib/searchhistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const regex = /^\s*\d+\s+(.*)$/;
class Searchhistory {
    constructor(nvim) {
        this.nvim = nvim;
        this.name = 'searchhistory';
        this.description = 'search history';
        this.defaultAction = 'feedkeys';
        this.actions = [];
        this.actions.push({
            name: 'feedkeys',
            execute: async (item) => {
                if (Array.isArray(item))
                    return;
                nvim.call('feedkeys', [`/${item.data.cmd}`, 'n'], true);
            }
        });
    }
    async loadItems(_context) {
        let { nvim } = this;
        let list = await nvim.eval(`split(execute("history search"),"\n")`);
        list = list.slice(1);
        let res = [];
        for (let line of list) {
            let ms = line.match(regex);
            if (ms) {
                let [, cmd] = ms;
                res.push({
                    label: cmd,
                    filterText: cmd,
                    data: {
                        cmd
                    }
                });
            }
        }
        return res;
    }
}
exports.default = Searchhistory;
